# BIOMART NOTES

### Ensembl GRCh37 ftp website, about text below: <http://grch37.ensembl.org/index.html>

**About this archive:**
- "This archive is based on Ensembl Release 75 data, and gives continuing access to human assembly GRCh37, as well as all our other release 75 species (data freeze March 2014) for comparative purposes. Human variation and regulation data has since been updated in March 2015"

### biomart docs: 
- .pdf of biomart documentation in projectHTML folder along with .html on RESTful_wget_access.html

### code from downloading using biomart website

```bash
8.12.19

project: comparing the different objects ENST point to, when ENST from GRCh37 matches GRCh38

FTP site for GRCh37 ensembl ftp://ftp.ensembl.org/pub/grch37/release-76/

# _______NOTES_______

Version of GRCh37 and GRCh38 file from v97


concern: if GRCh37 is frozen on ensembl75, but i use v97 of GRCh37 to compare...does this work? or should i use the version of 37 from ensembl75, in which case i will get way fewer UKB matched IDs

website of text below: http://grch37.ensembl.org/index.html

"About this archive
This archive is based on Ensembl Release 75 data, and gives continuing access to human assembly GRCh37, as well as all our other release 75 species (data freeze March 2014) for comparative purposes. Human variation and regulation data has since been updated in March 2015"


.pdf of biomart documentation in projectHTML folder along with .html on RESTful_wget_access.html



# _______using website_______
filename: biomartGRCh37p.13_v97_filterCCDSIDonly.txt
# xml of GRCh37p.13 v97 file downloaded, selected "Unique results only"
# 18,710 / 63,737 genes
# wc -l = 37,335
# http://grch37.ensembl.org/biomart/martview/f4c0f62fea09be6cf4a121710d2aa3bb

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE Query>
<Query  virtualSchemaName = "default" formatter = "TSV" header = "0" uniqueRows = "0" count = "" datasetConfigVersion = "0.6" >
			
	<Dataset name = "hsapiens_gene_ensembl" interface = "default" >
		<Filter name = "with_ccds" excluded = "0"/>
		<Attribute name = "ensembl_gene_id" />
		<Attribute name = "ensembl_gene_id_version" />
		<Attribute name = "ensembl_transcript_id" />
		<Attribute name = "ensembl_transcript_id_version" />
		<Attribute name = "ensembl_peptide_id" />
		<Attribute name = "ensembl_peptide_id_version" />
		<Attribute name = "ccds" />
		<Attribute name = "uniprotswissprot" />
		<Attribute name = "shares_cds_and_utr_with_ottt" />
	</Dataset>
</Query>
# ____________________________

filename: biomartGRCh37p.13_v97_filterHAVANAensemblTranscriptMatchOnly.txt
GRCH37 only difference is filter only for 
"With Transcript having exact match between ENSEMBL and HAVANA ID(s): Only"
genes 19,066/63,737 
wc -l = 28,773
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE Query>
<Query  virtualSchemaName = "default" formatter = "TSV" header = "0" uniqueRows = "0" count = "" datasetConfigVersion = "0.6" >
			
	<Dataset name = "hsapiens_gene_ensembl" interface = "default" >
		<Filter name = "with_shares_cds_and_utr_with_ottt" excluded = "0"/>
		<Attribute name = "ensembl_gene_id" />
		<Attribute name = "ensembl_gene_id_version" />
		<Attribute name = "ensembl_transcript_id" />
		<Attribute name = "ensembl_transcript_id_version" />
		<Attribute name = "ensembl_peptide_id" />
		<Attribute name = "ensembl_peptide_id_version" />
		<Attribute name = "ccds" />
		<Attribute name = "uniprotswissprot" />
		<Attribute name = "shares_cds_and_utr_with_ottt" />
	</Dataset>
</Query>
# _____________________
filename: biomartGRCh38p.12_v97_filterCCDSIDonly.txt
# xml of GRCh38p.12 v97 file downloaded, selected "Unique results only"
# no option to filter for "With Transcript having exact match between ENSEMBL and HAVANA ID(s): Only"
# filter instead for with_ccds
# 19,086/66,832 genes (filter with ccds id)
# wc -l = 42,992

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE Query>
<Query  virtualSchemaName = "default" formatter = "TSV" header = "0" uniqueRows = "0" count = "" datasetConfigVersion = "0.6" >
			
	<Dataset name = "hsapiens_gene_ensembl" interface = "default" >
		<Filter name = "with_ccds" excluded = "0"/>
		<Attribute name = "ensembl_gene_id" />
		<Attribute name = "ensembl_gene_id_version" />
		<Attribute name = "ensembl_transcript_id" />
		<Attribute name = "ensembl_transcript_id_version" />
		<Attribute name = "ensembl_peptide_id" />
		<Attribute name = "ensembl_peptide_id_version" />
		<Attribute name = "ccds" />
		<Attribute name = "uniprotswissprot" />
	</Dataset>
</Query>

# downloading the sequences of peptides:

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE Query>
<Query  virtualSchemaName = "default" formatter = "FASTA" header = "0" uniqueRows = "0" count = "" datasetConfigVersion = "0.6" >
				
	<Dataset name = "hsapiens_gene_ensembl" interface = "default" >
		<Filter name = "with_ccds" excluded = "0"/>
		<Filter name = "transcript_gencode_basic" excluded = "0"/>
		<Attribute name = "peptide" />
		<Attribute name = "ensembl_gene_id" />
		<Attribute name = "ensembl_gene_id_version" />
		<Attribute name = "ensembl_transcript_id" />
		<Attribute name = "ensembl_transcript_id_version" />
		<Attribute name = "ensembl_peptide_id" />
		<Attribute name = "ensembl_peptide_id_version" />
		<Attribute name = "uniprotswissprot" />
	</Dataset>
</Query>

xml of thing i need to download for both 37 and 38

http://grch37.ensembl.org/biomart/martview/4c6d8d4c400ea7671da67d954bc62328
    
    http://uswest.ensembl.org/biomart/martview/e26c8319ae734c2320d586230b1562ef


---
---
---

---
---
---

---
---
---
# first attempt with ENSG_ENST_ENSP key below, scroll to bottom of markdown for correct more recent code
---
---
---

---
---
---



# [1st attempt: TODO] 
1. load biomart downloaded files from ensembl release 97
2. simplify files to ENST, ENSG, and ENSP (both stable ID and version ID)
3. Join INNER on ENST, check for differences in ENSG and ENSP, also UKB ID

In [1]:
# packages
import os
import sys
import numpy as np
import pandas as pd
import csv

In [2]:
from IPython.display import display
pd.set_option('display.max_columns', None)
pd.options.display.max_seq_items = 2000

In [3]:
sys.path.append('/Users/mariapalafox/Desktop/Toolbox')
from all_funx import *

In [4]:
os.chdir("/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/BIOMART_ENSEMBL/")
print(os.listdir())

['v97sameStableIDGRCh37and38_diffENSPv_10.csv', 'TF_38_truelabeled_before_5361_keyid2.csv', '.DS_Store', 'TF38_22203.csv', 'TF_38_22202_keyid_v2.csv', 'NOTES.txt', 'GRCh37vsGRCh38_ensembl.sh', 'biomartGRCh38p.12_v97_filterCCDSIDonly.txt', 'v97StableIDmathchedGRCh38_duplicateStableID_fromUniqueUKBcrossRef_14.csv', 'entryUniProtCCDSset_simple.txt', 'StableIDs_allMatched_btw_GRCh37_38_ensemblv97_22202.csv', 'mapping_37.txt', 'biomart_CCDSIDonly_GRCh37_v97_peptideFasta.txt', 'StableIDs_ENST_ENSG_ENSP_v97ensembl_GRCh37_matchGRCh38_22193_fromBiomartCCDSonlyfilter.csv', 'UKBccdsEntry2stableENSP_fromTF38_stableIDkeyMatched_15745entries_counts2ENSPmultimapping.csv', 'TF37_22205.csv', 'Homo_sapiens.GRCh38.pep.all.fa', 'mapping_38.txt', 'v97versionDifferences_withSameStableID_GRCh37and38_22193.csv', 'TF_37_truelabeled_before_5361_keyid2.csv', 'biomartGRCh37p.13_v97_filterCCDSIDonly.txt', 'TF_37_22202_keyid_v2.csv', 'biomartGRCh37p.13_v97_filterHAVANAensemblTranscriptMatchOnly.txt', 'v97StableIDma

In [7]:
# files downloaded from biomart, specifics of file 
ccds37 = "biomartGRCh37p.13_v97_filterCCDSIDonly.txt"
ccds38 = "biomartGRCh38p.12_v97_filterCCDSIDonly.txt"
havana37 = "biomartGRCh37p.13_v97_filterHAVANAensemblTranscriptMatchOnly.txt"

# after dropping all nan values, ccds37 == hava37

In [8]:
ccds37 = pd.read_csv(ccds37,sep='\t')
ccds38 = pd.read_csv(ccds38,sep='\t')
hava37 = pd.read_csv(havana37,sep='\t')

In [28]:
describeMe(ccds37)
describeMe(ccds38)
describeMe(hava37)

(23334, 9)
Index(['Gene stable ID', 'Gene stable ID version', 'Transcript stable ID',
       'Transcript stable ID version', 'Protein stable ID',
       'Protein stable ID version', 'CCDS ID', 'UniProtKB/Swiss-Prot ID',
       'Transcript having exact match between ENSEMBL and HAVANA ID'],
      dtype='object')
Gene stable ID                                                 0
Gene stable ID version                                         0
Transcript stable ID                                           0
Transcript stable ID version                                   0
Protein stable ID                                              0
Protein stable ID version                                      0
CCDS ID                                                        0
UniProtKB/Swiss-Prot ID                                        0
Transcript having exact match between ENSEMBL and HAVANA ID    0
dtype: int64
(38719, 8)
Index(['Gene stable ID', 'Gene stable ID version', 'Transcript stable ID',
   

In [23]:
# dropping all NAN rows in all dataframes
ccds37.dropna(inplace=True)
ccds38.dropna(inplace=True)
hava37.dropna(inplace=True)

In [24]:
ccds37.reset_index(drop=True,inplace=True)
ccds38.reset_index(drop=True,inplace=True)
hava37.reset_index(drop=True,inplace=True)

In [25]:
ccds37.head(5)

Gene stable ID Gene stable ID version Transcript stable ID  \
0  ENSG00000215405      ENSG00000215405.3      ENST00000427390   
1  ENSG00000230031      ENSG00000230031.6      ENST00000454856   
2  ENSG00000138593      ENSG00000138593.4      ENST00000559471   
3  ENSG00000138593      ENSG00000138593.4      ENST00000261847   
4  ENSG00000166157     ENSG00000166157.12      ENST00000298232   

  Transcript stable ID version Protein stable ID Protein stable ID version  \
0            ENST00000427390.2   ENSP00000398615         ENSP00000398615.2   
1            ENST00000454856.4   ENSP00000456953         ENSP00000456953.1   
2            ENST00000559471.1   ENSP00000453854         ENSP00000453854.1   
3            ENST00000261847.3   ENSP00000261847         ENSP00000261847.3   
4            ENST00000298232.7   ENSP00000298232         ENSP00000298232.7   

     CCDS ID UniProtKB/Swiss-Prot ID  \
0  CCDS45184                  A8MZA4   
1  CCDS59248                  H3BUK9   
2  CCDS53942                  Q93073   
3  CCDS32234                  Q93073   
4  CCDS33512                  P56180   

  Transcript having exact match between ENSEMBL and HAVANA ID  
0                                 OTTHUMT00000414660           
1                                 OTTHUMT00000471435           
2                                 OTTHUMT00000417277           
3                                 OTTHUMT00000417278           
4                                 OTTHUMT00000157412

In [26]:
hava37.head(5)

Gene stable ID Gene stable ID version Transcript stable ID  \
0  ENSG00000215405      ENSG00000215405.3      ENST00000427390   
1  ENSG00000230031      ENSG00000230031.6      ENST00000454856   
2  ENSG00000138593      ENSG00000138593.4      ENST00000559471   
3  ENSG00000138593      ENSG00000138593.4      ENST00000261847   
4  ENSG00000166157     ENSG00000166157.12      ENST00000298232   

  Transcript stable ID version Protein stable ID Protein stable ID version  \
0            ENST00000427390.2   ENSP00000398615         ENSP00000398615.2   
1            ENST00000454856.4   ENSP00000456953         ENSP00000456953.1   
2            ENST00000559471.1   ENSP00000453854         ENSP00000453854.1   
3            ENST00000261847.3   ENSP00000261847         ENSP00000261847.3   
4            ENST00000298232.7   ENSP00000298232         ENSP00000298232.7   

     CCDS ID UniProtKB/Swiss-Prot ID  \
0  CCDS45184                  A8MZA4   
1  CCDS59248                  H3BUK9   
2  CCDS53942                  Q93073   
3  CCDS32234                  Q93073   
4  CCDS33512                  P56180   

  Transcript having exact match between ENSEMBL and HAVANA ID  
0                                 OTTHUMT00000414660           
1                                 OTTHUMT00000471435           
2                                 OTTHUMT00000417277           
3                                 OTTHUMT00000417278           
4                                 OTTHUMT00000157412

In [27]:
# are they the same 
ccds37.equals(hava37)

True

# ENSEMBL v97 GRCh37 df= 23,334
# ENSEMBL v97 GRCh38 df= 38,719

- how does this version compare with GENCODEv19 mapping to UKB, way more IDs to UKB
    - gencode website, although run by ensembl, is not maintained to updated mappings
    


# TODO: make files to comapre join with
1. only stable IDs (ENST joined inner, compare ENSG/ENSP/UKB) between assemblies
2. only versioned IDs (ENST again with same comparisons) i except more differences here
3. improt ccds files to compare ENST. 

In [31]:
print("37 columns: ", ccds37.columns)
print("38 columns: ", ccds38.columns)

37 columns:  Index(['Gene stable ID', 'Gene stable ID version', 'Transcript stable ID',
       'Transcript stable ID version', 'Protein stable ID',
       'Protein stable ID version', 'CCDS ID', 'UniProtKB/Swiss-Prot ID',
       'Transcript having exact match between ENSEMBL and HAVANA ID'],
      dtype='object')
38 columns:  Index(['Gene stable ID', 'Gene stable ID version', 'Transcript stable ID',
       'Transcript stable ID version', 'Protein stable ID',
       'Protein stable ID version', 'CCDS ID', 'UniProtKB/Swiss-Prot ID'],
      dtype='object')


In [32]:
# GRCh37 stable and versioned files created
G37stable = ccds37.filter(['Gene stable ID', 'Transcript stable ID', 'Protein stable ID', 'CCDS ID', 'UniProtKB/Swiss-Prot ID'], axis=1)
G37ver = ccds37.filter(['Gene stable ID version', 'Transcript stable ID version', 'Protein stable ID version', 'CCDS ID', 'UniProtKB/Swiss-Prot ID'], axis=1)
# GRCh38 stable and versioned files created
G38stable = ccds38.filter(['Gene stable ID', 'Transcript stable ID', 'Protein stable ID', 'CCDS ID', 'UniProtKB/Swiss-Prot ID'], axis=1)
G38ver = ccds38.filter(['Gene stable ID version', 'Transcript stable ID version', 'Protein stable ID version', 'CCDS ID', 'UniProtKB/Swiss-Prot ID'], axis=1)    

# TODO 1: compare stable ID file differences

In [33]:
G37stable.head(2)

Gene stable ID Transcript stable ID Protein stable ID    CCDS ID  \
0  ENSG00000215405      ENST00000427390   ENSP00000398615  CCDS45184   
1  ENSG00000230031      ENST00000454856   ENSP00000456953  CCDS59248   

  UniProtKB/Swiss-Prot ID  
0                  A8MZA4  
1                  H3BUK9

In [34]:
G38stable.head(2)

Gene stable ID Transcript stable ID Protein stable ID    CCDS ID  \
0  ENSG00000176555      ENST00000319988   ENSP00000321447  CCDS31488   
1  ENSG00000141682      ENST00000316660   ENSP00000326119  CCDS11975   

  UniProtKB/Swiss-Prot ID  
0                  Q8NGB4  
1                  Q13794

In [37]:
# gencode_myccds_genukb = pd.merge(gencodetx,myccds_genukb,how='inner',on=['ENST'])
stable37v38 = pd.merge(G37stable,G38stable,how='inner',on=['Transcript stable ID'])
describeMe(stable37v38)


(22414, 9)
Index(['Gene stable ID_x', 'Transcript stable ID', 'Protein stable ID_x',
       'CCDS ID_x', 'UniProtKB/Swiss-Prot ID_x', 'Gene stable ID_y',
       'Protein stable ID_y', 'CCDS ID_y', 'UniProtKB/Swiss-Prot ID_y'],
      dtype='object')
Gene stable ID_x             0
Transcript stable ID         0
Protein stable ID_x          0
CCDS ID_x                    0
UniProtKB/Swiss-Prot ID_x    0
Gene stable ID_y             0
Protein stable ID_y          0
CCDS ID_y                    0
UniProtKB/Swiss-Prot ID_y    0
dtype: int64


In [39]:
# change column order
cols = list(stable37v38.columns.values)

cols.pop(cols.index('Transcript stable ID')) # removes 'key_id' from list

stable37v38 = stable37v38.loc[:,['Transcript stable ID']+cols] # creates df w/ col in order

#stable37v38

Transcript stable ID Gene stable ID_x Protein stable ID_x  CCDS ID_x  \
0          ENST00000454856  ENSG00000230031     ENSP00000456953  CCDS59248   
1          ENST00000559471  ENSG00000138593     ENSP00000453854  CCDS53942   
2          ENST00000261847  ENSG00000138593     ENSP00000261847  CCDS32234   
3          ENST00000299443  ENSG00000166351     ENSP00000299443  CCDS13562   
4          ENST00000399848  ENSG00000168675     ENSP00000382741  CCDS32794   
5          ENST00000344577  ENSG00000188992     ENSP00000343331  CCDS13564   
6          ENST00000587757  ENSG00000168675     ENSP00000466178  CCDS32795   
7          ENST00000586765  ENSG00000168675     ENSP00000474783  CCDS42415   
8          ENST00000328795  ENSG00000183706     ENSP00000332500  CCDS32173   
9          ENST00000400577  ENSG00000185272     ENSP00000383421  CCDS46635   
10         ENST00000285667  ENSG00000155304     ENSP00000285667  CCDS13567   
11         ENST00000400566  ENSG00000155307     ENSP00000383411  CCDS42906   
12         ENST00000388901  ENSG00000166200     ENSP00000373553  CCDS32235   
13         ENST00000299259  ENSG00000166200     ENSP00000299259  CCDS45257   
14         ENST00000338376  ENSG00000137824     ENSP00000342493  CCDS10063   
15         ENST00000260447  ENSG00000137880     ENSP00000260447  CCDS10064   
16         ENST00000322247  ENSG00000177150     ENSP00000323635  CCDS11866   
17         ENST00000220496  ENSG00000104129     ENSP00000220496  CCDS10065   
18         ENST00000285679  ENSG00000155313     ENSP00000285679  CCDS33515   
19         ENST00000285681  ENSG00000155313     ENSP00000285681  CCDS63337   
20         ENST00000315087  ENSG00000101638     ENSP00000321343  CCDS11930   
21         ENST00000284878  ENSG00000154639     ENSP00000284878  CCDS33519   
22         ENST00000400166  ENSG00000154639     ENSP00000383030  CCDS56205   
23         ENST00000356275  ENSG00000154639     ENSP00000348620  CCDS56207   
24         ENST00000400165  ENSG00000154639     ENSP00000383029  CCDS56206   
25         ENST00000543302  ENSG00000101654     ENSP00000446426  CCDS11867   
26         ENST00000585916  ENSG00000078043     ENSP00000465676  CCDS32824   
27         ENST00000324794  ENSG00000078043     ENSP00000317163  CCDS32825   
28         ENST00000339775  ENSG00000154640     ENSP00000344609  CCDS46636   
29         ENST00000348354  ENSG00000154640     ENSP00000284879  CCDS13569   
...                    ...              ...                 ...        ...   
22384      ENST00000367279  ENSG00000143851     ENSP00000356248   CCDS1422   
22385      ENST00000318348  ENSG00000139433     ENSP00000315263   CCDS9136   
22386      ENST00000367278  ENSG00000133067     ENSP00000356247  CCDS30971   
22387      ENST00000255432  ENSG00000133067     ENSP00000255432   CCDS1424   
22388      ENST00000439764  ENSG00000133067     ENSP00000387869  CCDS30972   
22389      ENST00000376663  ENSG00000168283     ENSP00000365851   CCDS7138   
22390      ENST00000356764  ENSG00000077157     ENSP00000349206  CCDS53459   
22391      ENST00000405876  ENSG00000139437     ENSP00000384520   CCDS9137   
22392      ENST00000312777  ENSG00000139437     ENSP00000324404   CCDS9137   
22393      ENST00000367267  ENSG00000143858     ENSP00000356236   CCDS1427   
22394      ENST00000367268  ENSG00000143858     ENSP00000356237   CCDS1427   
22395      ENST00000586346  ENSG00000181513     ENSP00000465484  CCDS45710   
22396      ENST00000398322  ENSG00000181513     ENSP00000381367  CCDS42348   
22397      ENST00000376955  ENSG00000181513     ENSP00000366154  CCDS45711   
22398      ENST00000367265  ENSG00000117139     ENSP00000356234  CCDS30974   
22399      ENST00000367262  ENSG00000183155     ENSP00000356231   CCDS1428   
22400      ENST00000367261  ENSG00000117153     ENSP00000356230   CCDS1429   
22401      ENST00000340990  ENSG00000159346     ENSP00000341785   CCDS1430   
22402      ENST00000367249  ENSG00000159348     ENSP00000356218   CCDS1431   
22403      ENST00000367242  

### joined on ENST merged =(22,414- 9)
#### X = 37
#### Y = 38 
#### inner: use intersection of keys from both frames, similar to a SQL inner join; preserve the order of the left keys.

# same transcript stable ID (22,414 rows):
- how many rows have different Gene stable ID =
```
0            match  22380
1        different     34
```
- how many rows have diff Protein stable ID =
```
0               match  22413
1           different      1
```
- how many rows have diff CCDS ID =
```
0            match  22311
1        different    103
```
- how many rows have diff UKB ID =
```
0           match  22328
1       different     86
```

In [43]:
# gene column compare
xx="Gene stable ID_x"
yy="Gene stable ID_y"

conditions = [
    stable37v38[xx] == stable37v38[yy], 
    stable37v38[xx] != stable37v38[yy]]
choices = ["match", "different"]

stable37v38['geneStable_check'] = np.select(conditions, choices, default=np.nan)

In [44]:
checkColumnValues(stable37v38,'geneStable_check')

  geneStable_check  Count
0            match  22380
1        different     34


In [45]:
# protein compare
xx="Protein stable ID_x"
yy="Protein stable ID_y"

conditions = [
    stable37v38[xx] == stable37v38[yy], 
    stable37v38[xx] != stable37v38[yy]]
choices = ["match", "different"]

stable37v38['proteinStable_check'] = np.select(conditions, choices, default=np.nan)

In [46]:
checkColumnValues(stable37v38,'proteinStable_check')

  proteinStable_check  Count
0               match  22413
1           different      1


In [47]:
xx="UniProtKB/Swiss-Prot ID_x"
yy="UniProtKB/Swiss-Prot ID_y"

conditions = [
    stable37v38[xx] == stable37v38[yy], 
    stable37v38[xx] != stable37v38[yy]]
choices = ["match", "different"]

stable37v38['ukbStable_check'] = np.select(conditions, choices, default=np.nan)

In [48]:
checkColumnValues(stable37v38,'ukbStable_check')

  ukbStable_check  Count
0           match  22328
1       different     86


In [49]:
# CCDS compare

xx="CCDS ID_x"
yy="CCDS ID_y"

conditions = [
    stable37v38[xx] == stable37v38[yy], 
    stable37v38[xx] != stable37v38[yy]]
choices = ["match", "different"]

stable37v38['ccdsStable_check'] = np.select(conditions, choices, default=np.nan)

In [50]:
checkColumnValues(stable37v38,'ccdsStable_check')

  ccdsStable_check  Count
0            match  22311
1        different    103


# TODO 2: compare versioned ID file differences

In [51]:
G37ver.head(2)

Gene stable ID version Transcript stable ID version  \
0      ENSG00000215405.3            ENST00000427390.2   
1      ENSG00000230031.6            ENST00000454856.4   

  Protein stable ID version    CCDS ID UniProtKB/Swiss-Prot ID  
0         ENSP00000398615.2  CCDS45184                  A8MZA4  
1         ENSP00000456953.1  CCDS59248                  H3BUK9

In [52]:
G38ver.head(2)

Gene stable ID version Transcript stable ID version  \
0      ENSG00000176555.1            ENST00000319988.1   
1     ENSG00000141682.11            ENST00000316660.6   

  Protein stable ID version    CCDS ID UniProtKB/Swiss-Prot ID  
0         ENSP00000321447.1  CCDS31488                  Q8NGB4  
1         ENSP00000326119.6  CCDS11975                  Q13794

In [53]:
ver37v38 = pd.merge(G37ver,G38ver,how='inner',on=['Transcript stable ID version'])
describeMe(ver37v38)

(2891, 9)
Index(['Gene stable ID version_x', 'Transcript stable ID version',
       'Protein stable ID version_x', 'CCDS ID_x', 'UniProtKB/Swiss-Prot ID_x',
       'Gene stable ID version_y', 'Protein stable ID version_y', 'CCDS ID_y',
       'UniProtKB/Swiss-Prot ID_y'],
      dtype='object')
Gene stable ID version_x        0
Transcript stable ID version    0
Protein stable ID version_x     0
CCDS ID_x                       0
UniProtKB/Swiss-Prot ID_x       0
Gene stable ID version_y        0
Protein stable ID version_y     0
CCDS ID_y                       0
UniProtKB/Swiss-Prot ID_y       0
dtype: int64


# MATCHED on version ID only have 2,891 matching between 37 and 38

---
---
---

# creating match all file (37 = 38 for gene/tx/pro stable IDs & UKB/CCDS)
1. map gene name to file to see how many genes are unique
2. map CCDS to file to see how many sequences are different


In [55]:
describeMe(stable37v38)

(22414, 13)
Index(['Transcript stable ID', 'Gene stable ID_x', 'Protein stable ID_x',
       'CCDS ID_x', 'UniProtKB/Swiss-Prot ID_x', 'Gene stable ID_y',
       'Protein stable ID_y', 'CCDS ID_y', 'UniProtKB/Swiss-Prot ID_y',
       'geneStable_check', 'proteinStable_check', 'ukbStable_check',
       'ccdsStable_check'],
      dtype='object')
Transcript stable ID         0
Gene stable ID_x             0
Protein stable ID_x          0
CCDS ID_x                    0
UniProtKB/Swiss-Prot ID_x    0
Gene stable ID_y             0
Protein stable ID_y          0
CCDS ID_y                    0
UniProtKB/Swiss-Prot ID_y    0
geneStable_check             0
proteinStable_check          0
ukbStable_check              0
ccdsStable_check             0
dtype: int64


# from biomart 37 and 38 download originally
```bash
#!/bin/bash

# GRCh37vsGRCh38.sh
# 8.12.19
# mfpfox

# more comments @ https://www.biostars.org/p/257665/
# To create this search
# wget -O result.txt 'http://www.ensembl.org/biomart/martservice?query=
# then go to biomart website and search, then save as xml file
# the xml is formatted into 1 line and added as your query=""

# I THINK THE VERSION OF ENSEMBL FOR BOTH ASSEMBLY FILES IS V97, SINCE THAT IS CURRENT AND I DID NOT SPECIFY A VERSION. NEED TO CHECK HOW DIFFERENT VERSION 76 IS....
# WEBSITE FOR ENSEMBL QUERY: http://grch37.ensembl.org/index.html
# MORE INFO IN NOTES.SH

# bash script gives you a list of ENSTs mapping to different ENSGs in grch37 and grch38:

#download of all ENST mapped to ENSG from Ensembls grch37 biomart
wget -O mapping_37.txt 'http://grch37.ensembl.org/biomart/martservice?query=<Query virtualSchemaName="default" formatter="TSV" header="0" uniqueRows="1" count="" datasetConfigVersion="0.6" completionStamp="1"><Dataset name="hsapiens_gene_ensembl" interface="default"><Attribute name="ensembl_transcript_id"/><Attribute name="ensembl_gene_id"/><Attribute name = "hgnc_symbol" /><Attribute name = "uniprotswissprot" /></Dataset></Query>'

#download of all ENST mapped to ENSG from Ensembls grch38 biomart
wget -O mapping_38.txt 'http://www.ensembl.org/biomart/martservice?query=<Query virtualSchemaName="default" formatter="TSV" header="0" uniqueRows="1" count="" datasetConfigVersion="0.6" completionStamp="1"><Dataset name="hsapiens_gene_ensembl" interface="default"><Attribute name="ensembl_transcript_id"/><Attribute name="ensembl_gene_id" /><Attribute name = "hgnc_symbol" /><Attribute name = "uniprotswissprot" /></Dataset></Query>'
```
# goal is quantify difference in all ensembl TRANSCRIPTS, since i already quantified difference in the set of transcripts that have a CCDS ID



In [60]:
# all transcripts downloaded in first biomart query
all37 = pd.read_csv("mapping_37.txt",sep='\t',header=None)
all38 = pd.read_csv("mapping_38.txt",sep='\t',header=None)

In [64]:
# changing columns
all37.columns=['ENST','ENSG','HGNC_symbol','UKB']
all37.head(2)

ENST             ENSG HGNC_symbol  UKB
0  ENST00000566782  ENSG00000261657    SLC25A26  NaN
1  ENST00000562780  ENSG00000261657    SLC25A26  NaN

In [65]:
all38.columns=['ENST','ENSG','HGNC_symbol','UKB']
all38.head(2)

ENST             ENSG HGNC_symbol     UKB
0  ENST00000361390  ENSG00000198888      MT-ND1  P03886
1  ENST00000361453  ENSG00000198763      MT-ND2  P03891

In [68]:
describeMe(all37)
describeMe(all38)

(45864, 4)
Index(['ENST', 'ENSG', 'HGNC_symbol', 'UKB'], dtype='object')
ENST           0
ENSG           0
HGNC_symbol    0
UKB            0
dtype: int64
(47110, 4)
Index(['ENST', 'ENSG', 'HGNC_symbol', 'UKB'], dtype='object')
ENST           0
ENSG           0
HGNC_symbol    0
UKB            0
dtype: int64


In [67]:
# drop all na values
all37.dropna(inplace=True)
all38.dropna(inplace=True)
all37.reset_index(drop=True,inplace=True)
all38.reset_index(drop=True,inplace=True)

In [69]:
allstable37v38 = pd.merge(all37,all38,how='inner',on=['ENST'])
describeMe(allstable37v38)

(39094, 7)
Index(['ENST', 'ENSG_x', 'HGNC_symbol_x', 'UKB_x', 'ENSG_y', 'HGNC_symbol_y',
       'UKB_y'],
      dtype='object')
ENST             0
ENSG_x           0
HGNC_symbol_x    0
UKB_x            0
ENSG_y           0
HGNC_symbol_y    0
UKB_y            0
dtype: int64


# merged all biomart 37 and 38 on ENST again, total rows 39,094 ( dropped all NAN values before merge )

# - how many rows have different Gene stable ID =
```
0      match  39000
1  different     94
```
# - how many rows have diff UKB ID =
```
0      match  38537
1  different    557
```
# - how many rows have diff HGNC_symbol ID =
```
0      match  37582
1  different   1512
```

In [70]:
# ENSG compare

xx="ENSG_x"
yy="ENSG_y"

conditions = [
    allstable37v38[xx] == allstable37v38[yy], 
    allstable37v38[xx] != allstable37v38[yy]]
choices = ["match", "different"]

allstable37v38['ENSG_check'] = np.select(conditions, choices, default=np.nan)

In [71]:
checkColumnValues(allstable37v38,'ENSG_check')

  ENSG_check  Count
0      match  39000
1  different     94


In [72]:
# HGNC compare

xx="HGNC_symbol_x"
yy="HGNC_symbol_y"

conditions = [
    allstable37v38[xx] == allstable37v38[yy], 
    allstable37v38[xx] != allstable37v38[yy]]
choices = ["match", "different"]

allstable37v38['HGNC_check'] = np.select(conditions, choices, default=np.nan)

In [73]:
checkColumnValues(allstable37v38,'HGNC_check')

  HGNC_check  Count
0      match  37582
1  different   1512


In [74]:
# UKB compare

xx="UKB_x"
yy="UKB_y"

conditions = [
    allstable37v38[xx] == allstable37v38[yy], 
    allstable37v38[xx] != allstable37v38[yy]]
choices = ["match", "different"]

allstable37v38['UKB_check'] = np.select(conditions, choices, default=np.nan)

In [75]:
checkColumnValues(allstable37v38,'UKB_check')

   UKB_check  Count
0      match  38537
1  different    557


In [77]:
# creating file with just ENST and HGNC, 
# only rows that matched between version, 
# then mapping this to other merged files from CCDS only download

# simpleHGNCmatch = allstable37v38.loc[(allstable37v38['HGNC_check'] == 'match')]
# onlyHGNC_ENST = simpleHGNCmatch.filter(['ENST','HGNC_symbol_x'])
# describeMe(onlyHGNC_ENST)

(37582, 2)
Index(['ENST', 'HGNC_symbol_x'], dtype='object')
ENST             0
HGNC_symbol_x    0
dtype: int64


In [82]:
#enstUnique= onlyHGNC_ENST['ENST'].unique()

# adding gene symbol column to the CCDS ID only download merge file

In [80]:
# describeMe(stable37v38)

(22414, 13)
Index(['ENST', 'Gene stable ID_x', 'Protein stable ID_x', 'CCDS ID_x',
       'UniProtKB/Swiss-Prot ID_x', 'Gene stable ID_y', 'Protein stable ID_y',
       'CCDS ID_y', 'UniProtKB/Swiss-Prot ID_y', 'geneStable_check',
       'proteinStable_check', 'ukbStable_check', 'ccdsStable_check'],
      dtype='object')
ENST                         0
Gene stable ID_x             0
Protein stable ID_x          0
CCDS ID_x                    0
UniProtKB/Swiss-Prot ID_x    0
Gene stable ID_y             0
Protein stable ID_y          0
CCDS ID_y                    0
UniProtKB/Swiss-Prot ID_y    0
geneStable_check             0
proteinStable_check          0
ukbStable_check              0
ccdsStable_check             0
dtype: int64


In [79]:
# change col name to ENST
stable37v38.columns = ['ENST', 'Gene stable ID_x', 'Protein stable ID_x',
       'CCDS ID_x', 'UniProtKB/Swiss-Prot ID_x', 'Gene stable ID_y',
       'Protein stable ID_y', 'CCDS ID_y', 'UniProtKB/Swiss-Prot ID_y',
       'geneStable_check', 'proteinStable_check', 'ukbStable_check',
       'ccdsStable_check']

In [81]:
# now join this file with the one I made from CCDS only merge
hgnc_stable37v38 = pd.merge(stable37v38,onlyHGNC_ENST,how='inner',on=['ENST'])
describeMe(hgnc_stable37v38)

(22953, 14)
Index(['ENST', 'Gene stable ID_x', 'Protein stable ID_x', 'CCDS ID_x',
       'UniProtKB/Swiss-Prot ID_x', 'Gene stable ID_y', 'Protein stable ID_y',
       'CCDS ID_y', 'UniProtKB/Swiss-Prot ID_y', 'geneStable_check',
       'proteinStable_check', 'ukbStable_check', 'ccdsStable_check',
       'HGNC_symbol_x'],
      dtype='object')
ENST                         0
Gene stable ID_x             0
Protein stable ID_x          0
CCDS ID_x                    0
UniProtKB/Swiss-Prot ID_x    0
Gene stable ID_y             0
Protein stable ID_y          0
CCDS ID_y                    0
UniProtKB/Swiss-Prot ID_y    0
geneStable_check             0
proteinStable_check          0
ukbStable_check              0
ccdsStable_check             0
HGNC_symbol_x                0
dtype: int64


# NOT BEST WAY TO ADD GENE SYMBOL BACK.
## NOW GOING TO DROP ALL ROWS IN CCDS ONLY FILE THAT HAD MISSMATCHES BETWEEN STABLE IDS AND SAVE THIS FILE


In [85]:
describeMe(stable37v38)

(22414, 13)
Index(['ENST', 'Gene stable ID_x', 'Protein stable ID_x', 'CCDS ID_x',
       'UniProtKB/Swiss-Prot ID_x', 'Gene stable ID_y', 'Protein stable ID_y',
       'CCDS ID_y', 'UniProtKB/Swiss-Prot ID_y', 'geneStable_check',
       'proteinStable_check', 'ukbStable_check', 'ccdsStable_check'],
      dtype='object')
ENST                         0
Gene stable ID_x             0
Protein stable ID_x          0
CCDS ID_x                    0
UniProtKB/Swiss-Prot ID_x    0
Gene stable ID_y             0
Protein stable ID_y          0
CCDS ID_y                    0
UniProtKB/Swiss-Prot ID_y    0
geneStable_check             0
proteinStable_check          0
ukbStable_check              0
ccdsStable_check             0
dtype: int64


In [92]:
# drop all with different values

checkMatch = stable37v38[~(stable37v38[['geneStable_check','proteinStable_check','ukbStable_check','ccdsStable_check']] == 'different').any(axis=1)]
checkMatch.reset_index(drop=True,inplace=True)

describeMe(checkMatch)

(22202, 13)
Index(['ENST', 'Gene stable ID_x', 'Protein stable ID_x', 'CCDS ID_x',
       'UniProtKB/Swiss-Prot ID_x', 'Gene stable ID_y', 'Protein stable ID_y',
       'CCDS ID_y', 'UniProtKB/Swiss-Prot ID_y', 'geneStable_check',
       'proteinStable_check', 'ukbStable_check', 'ccdsStable_check'],
      dtype='object')
ENST                         0
Gene stable ID_x             0
Protein stable ID_x          0
CCDS ID_x                    0
UniProtKB/Swiss-Prot ID_x    0
Gene stable ID_y             0
Protein stable ID_y          0
CCDS ID_y                    0
UniProtKB/Swiss-Prot ID_y    0
geneStable_check             0
proteinStable_check          0
ukbStable_check              0
ccdsStable_check             0
dtype: int64


In [93]:
len(checkMatch['ENST'].unique())

# 12 rows with redundancy in ENST

22193

In [97]:
# all rows with duplicate ENST
dupes = checkMatch[checkMatch.duplicated(['ENST'], keep= False)]
dupes

ENST Gene stable ID_x Protein stable ID_x  CCDS ID_x  \
4222   ENST00000529639  ENSG00000149806     ENSP00000435370   CCDS8095   
4223   ENST00000529639  ENSG00000149806     ENSP00000435370   CCDS8095   
17089  ENST00000360004  ENSG00000196126     ENSP00000353099  CCDS47409   
17090  ENST00000360004  ENSG00000196126     ENSP00000353099  CCDS47409   
17091  ENST00000360004  ENSG00000196126     ENSP00000353099  CCDS47409   
17092  ENST00000360004  ENSG00000196126     ENSP00000353099  CCDS47409   
20221  ENST00000412585  ENSG00000234745     ENSP00000399168  CCDS34394   
20222  ENST00000412585  ENSG00000234745     ENSP00000399168  CCDS34394   
20223  ENST00000412585  ENSG00000234745     ENSP00000399168  CCDS34394   
20224  ENST00000412585  ENSG00000234745     ENSP00000399168  CCDS34394   
20225  ENST00000412585  ENSG00000234745     ENSP00000399168  CCDS34394   
20226  ENST00000412585  ENSG00000234745     ENSP00000399168  CCDS34394   

      UniProtKB/Swiss-Prot ID_x Gene stable ID_y Protein stable ID_y  \
4222                     P35544  ENSG00000149806     ENSP00000435370   
4223                     P62861  ENSG00000149806     ENSP00000435370   
17089                    Q9GIY3  ENSG00000196126     ENSP00000353099   
17090                    P04229  ENSG00000196126     ENSP00000353099   
17091                    Q29974  ENSG00000196126     ENSP00000353099   
17092                    P01911  ENSG00000196126     ENSP00000353099   
20221                    Q31612  ENSG00000234745     ENSP00000399168   
20222                    P01889  ENSG00000234745     ENSP00000399168   
20223                    P30480  ENSG00000234745     ENSP00000399168   
20224                    Q29836  ENSG00000234745     ENSP00000399168   
20225                    Q31610  ENSG00000234745     ENSP00000399168   
20226                    P30486  ENSG00000234745     ENSP00000399168   

       CCDS ID_y UniProtKB/Swiss-Prot ID_y geneStable_check  \
4222    CCDS8095                    P35544            match   
4223    CCDS8095                    P62861            match   
17089  CCDS47409                    Q9GIY3            match   
17090  CCDS47409                    P04229            match   
17091  CCDS47409                    Q29974            match   
17092  CCDS47409                    P01911            match   
20221  CCDS34394                    Q31612            match   
20222  CCDS34394                    P01889            match   
20223  CCDS34394                    P30480            match   
20224  CCDS34394                    Q29836            match   
20225  CCDS34394                    Q31610            match   
20226  CCDS34394                    P30486            match   

      proteinStable_check ukbStable_check ccdsStable_check  
4222                match           match            match  
4223                match           match            match  
17089               match           match            match  
17090               match           match            match  
17091               match           match            match  
17092               match           match            match  
20221               match           match            match  
20222               match           match            match  
20223               match           match            match  
20224               match           match            match  
20225               match           match            match  
20226               match           match            match

In [98]:
checkMatch.to_csv("StableIDs_allMatched_btw_GRCh37_38_ensemblv97_22202.csv",index=False)

In [4]:
checkMatch = pd.read_csv("StableIDs_allMatched_btw_GRCh37_38_ensemblv97_22202.csv")

In [6]:
checksimple = checkMatch.filter(['ENST','Gene stable ID_x','Protein stable ID_x'])

In [9]:
checksimple.shape

(22202, 3)

In [10]:
nodups = checksimple.drop_duplicates()
nodups.shape

(22193, 3)

In [11]:
nodups.reset_index(drop=True,inplace=True)

In [12]:
nodups.to_csv("StableIDs_ENST_ENSG_ENSP_v97ensembl_GRCh37_matchGRCh38_22193_fromBiomartCCDSonlyfilter.csv",index=False)

---
---
---
---
---
---

---
---
---
---
---
---

# 2nd corrent attempt code below 9.4.19: 

---
---
---
---
---
---

---
---
---
---
---
---


# Creating new keyID 'ENSG_ENST_ENSP_UKBSP_CCDS'
#### this should help problem of duplicates in 38 and 37 caused by multi mapping of ENSG_ENST_ENSP to unique UKB IDs

```bash
# review
X = 37

Y = 38

inner merge on ENST stable, compared ENSG/ENSP/UKBswissProt/CCDS all stable for v97 ensembl downloaded from biomart
```
9.4.19

mfpfox



In [5]:
checkMatch = pd.read_csv("StableIDs_allMatched_btw_GRCh37_38_ensemblv97_22202.csv")

In [6]:
checkMatch

ENST Gene stable ID_x Protein stable ID_x  CCDS ID_x  \
0      ENST00000454856  ENSG00000230031     ENSP00000456953  CCDS59248   
1      ENST00000559471  ENSG00000138593     ENSP00000453854  CCDS53942   
2      ENST00000261847  ENSG00000138593     ENSP00000261847  CCDS32234   
3      ENST00000299443  ENSG00000166351     ENSP00000299443  CCDS13562   
4      ENST00000399848  ENSG00000168675     ENSP00000382741  CCDS32794   
5      ENST00000344577  ENSG00000188992     ENSP00000343331  CCDS13564   
6      ENST00000587757  ENSG00000168675     ENSP00000466178  CCDS32795   
7      ENST00000586765  ENSG00000168675     ENSP00000474783  CCDS42415   
8      ENST00000328795  ENSG00000183706     ENSP00000332500  CCDS32173   
9      ENST00000400577  ENSG00000185272     ENSP00000383421  CCDS46635   
10     ENST00000285667  ENSG00000155304     ENSP00000285667  CCDS13567   
11     ENST00000400566  ENSG00000155307     ENSP00000383411  CCDS42906   
12     ENST00000388901  ENSG00000166200     ENSP00000373553  CCDS32235   
13     ENST00000299259  ENSG00000166200     ENSP00000299259  CCDS45257   
14     ENST00000338376  ENSG00000137824     ENSP00000342493  CCDS10063   
15     ENST00000260447  ENSG00000137880     ENSP00000260447  CCDS10064   
16     ENST00000322247  ENSG00000177150     ENSP00000323635  CCDS11866   
17     ENST00000220496  ENSG00000104129     ENSP00000220496  CCDS10065   
18     ENST00000285679  ENSG00000155313     ENSP00000285679  CCDS33515   
19     ENST00000285681  ENSG00000155313     ENSP00000285681  CCDS63337   
20     ENST00000315087  ENSG00000101638     ENSP00000321343  CCDS11930   
21     ENST00000284878  ENSG00000154639     ENSP00000284878  CCDS33519   
22     ENST00000400166  ENSG00000154639     ENSP00000383030  CCDS56205   
23     ENST00000356275  ENSG00000154639     ENSP00000348620  CCDS56207   
24     ENST00000400165  ENSG00000154639     ENSP00000383029  CCDS56206   
25     ENST00000543302  ENSG00000101654     ENSP00000446426  CCDS11867   
26     ENST00000585916  ENSG00000078043     ENSP00000465676  CCDS32824   
27     ENST00000324794  ENSG00000078043     ENSP00000317163  CCDS32825   
28     ENST00000339775  ENSG00000154640     ENSP00000344609  CCDS46636   
29     ENST00000348354  ENSG00000154640     ENSP00000284879  CCDS13569   
...                ...              ...                 ...        ...   
22172  ENST00000367279  ENSG00000143851     ENSP00000356248   CCDS1422   
22173  ENST00000318348  ENSG00000139433     ENSP00000315263   CCDS9136   
22174  ENST00000367278  ENSG00000133067     ENSP00000356247  CCDS30971   
22175  ENST00000255432  ENSG00000133067     ENSP00000255432   CCDS1424   
22176  ENST00000439764  ENSG00000133067     ENSP00000387869  CCDS30972   
22177  ENST00000376663  ENSG00000168283     ENSP00000365851   CCDS7138   
22178  ENST00000356764  ENSG00000077157     ENSP00000349206  CCDS53459   
22179  ENST00000405876  ENSG00000139437     ENSP00000384520   CCDS9137   
22180  ENST00000312777  ENSG00000139437     ENSP00000324404   CCDS9137   
22181  ENST00000367267  ENSG00000143858     ENSP00000356236   CCDS1427   
22182  ENST00000367268  ENSG00000143858     ENSP00000356237   CCDS1427   
22183  ENST00000586346  ENSG00000181513     ENSP00000465484  CCDS45710   
22184  ENST00000398322  ENSG00000181513     ENSP00000381367  CCDS42348   
22185  ENST00000376955  ENSG00000181513     ENSP00000366154  CCDS45711   
22186  ENST00000367265  ENSG00000117139     ENSP00000356234  CCDS30974   
22187  ENST00000367262  ENSG00000183155     ENSP00000356231   CCDS1428   
22188  ENST00000367261  ENSG00000117153     ENSP00000356230   CCDS1429   
22189  ENST00000340990  ENSG00000159346     ENSP00000341785   CCDS1430   
22190  ENST00000367249  ENSG00000159348     ENSP00000356218   CCDS1431   
22191  ENST00000367242  ENSG00000163444     ENSP00000356211   CCDS1432   
22192  ENST00000355312  ENSG00000139436     ENSP00000347464   CCDS9138   
22193  ENST00000361006  ENSG00000139436     ENSP00000354282  CCDS44968   
22194  ENST00000553118

In [7]:
checkMatch.columns

Index(['ENST', 'Gene stable ID_x', 'Protein stable ID_x', 'CCDS ID_x',
       'UniProtKB/Swiss-Prot ID_x', 'Gene stable ID_y', 'Protein stable ID_y',
       'CCDS ID_y', 'UniProtKB/Swiss-Prot ID_y', 'geneStable_check',
       'proteinStable_check', 'ukbStable_check', 'ccdsStable_check'],
      dtype='object')

In [8]:
# create stable_id for mapping to stableMatch IDs to versioned rows in 37 and 38 file 
# extension of original key
# does not matter than some have _x and some have _y, they are identical

checkMatch['stable_id2'] = checkMatch['Gene stable ID_x'].astype(str) + \
'_' + checkMatch['ENST'].astype(str) + \
'_' + checkMatch['Protein stable ID_y'].astype(str) + \
'_' + checkMatch['CCDS ID_y'] + \
'_' + checkMatch['UniProtKB/Swiss-Prot ID_y']


In [9]:
checkMatch['stable_id2'][4]

'ENSG00000168675_ENST00000399848_ENSP00000382741_CCDS32794_O15165'

# TODO-0 import biomart v97 download files and filter based on new KeyID {ENSG_ENST_ENSP_UKBSP_CCDS}

```python
# files downloaded from biomart, specifics of file 
# used in original keyID experiment from 19_08_20_DEPENDENTvsINDEPENDENT...

# importing file
ccds37 = "biomartGRCh37p.13_v97_filterCCDSIDonly.txt"
ccds38 = "biomartGRCh38p.12_v97_filterCCDSIDonly.txt"

version37 = pd.read_csv(ccds37,sep='\t')
version38 = pd.read_csv(ccds38,sep='\t')

# drop Nan values reset index
version37.dropna(inplace=True)
version38.dropna(inplace=True)
# reset index
version37.reset_index(drop=True,inplace=True)
version38.reset_index(drop=True,inplace=True)

describeMe(version37)
describeMe(version38)
describeMe(stable)

# create keyID
# GRCh37 v97 biomart
version37['stable_id2'] = version37['Gene stable ID'].astype(str) + \
'_' + version37['Transcript stable ID'].astype(str) + \
'_' + version37['Protein stable ID'].astype(str) + \
'_' + version37['CCDS ID'] + \
'_' + version37['UniProtKB/Swiss-Prot ID']
# GRCh38 v97 biomart
version38['stable_id2'] = version38['Gene stable ID'].astype(str) + \
'_' + version38['Transcript stable ID'].astype(str) + \
'_' + version38['Protein stable ID'].astype(str) + \
'_' + version38['CCDS ID'] + \
'_' + version38['UniProtKB/Swiss-Prot ID']

# get total # of unique UKB IDs
print(len(checkMatch.stable_id2.unique()))
print(len(version37.stable_id2.unique()))
print(len(version38.stable_id2.unique()))


# label each 37 and 38 version file with True False 
# if stable_id is in the unique set of ID from my stable ID overlap between assemblies analysis
def addcolumnconditional(mapList, df, dfcol, newcol): 
    mendel = []
    for g in df[dfcol]:
        if g in mapList: 
            mendel.append("True")
        else:
            mendel.append("False")
    df.loc[:,newcol] = mendel
    return df


TF37 = addcolumnconditional(stableList, version37, "stable_id", "37match38stable_id")

# drop all rows that have FALSE for match 
TF37.drop(TF37[TF37["37match38stable_id"] == "False"].index, inplace = True)
TF38.drop(TF38[TF38["37match38stable_id"] == "False"].index, inplace = True)

checkColumnValues(TF37, "37match38stable_id")
# same for 38

# save these files
```

In [10]:
# files downloaded from biomart, specifics of file 
# used in original keyID experiment from 19_08_20_DEPENDENTvsINDEPENDENT...

# importing file
ccds37 = "biomartGRCh37p.13_v97_filterCCDSIDonly.txt"
ccds38 = "biomartGRCh38p.12_v97_filterCCDSIDonly.txt"

version37 = pd.read_csv(ccds37,sep='\t')
version38 = pd.read_csv(ccds38,sep='\t')

# drop Nan values reset index
version37.dropna(inplace=True)
version38.dropna(inplace=True)
# reset index
version37.reset_index(drop=True,inplace=True)
version38.reset_index(drop=True,inplace=True)

In [13]:

describeMe(version37)
describeMe(version38)
describeMe(checkMatch)

(23334, 9)
Index(['Gene stable ID', 'Gene stable ID version', 'Transcript stable ID',
       'Transcript stable ID version', 'Protein stable ID',
       'Protein stable ID version', 'CCDS ID', 'UniProtKB/Swiss-Prot ID',
       'Transcript having exact match between ENSEMBL and HAVANA ID'],
      dtype='object')
Gene stable ID                                                 0
Gene stable ID version                                         0
Transcript stable ID                                           0
Transcript stable ID version                                   0
Protein stable ID                                              0
Protein stable ID version                                      0
CCDS ID                                                        0
UniProtKB/Swiss-Prot ID                                        0
Transcript having exact match between ENSEMBL and HAVANA ID    0
dtype: int64
(38719, 8)
Index(['Gene stable ID', 'Gene stable ID version', 'Transcript stable ID',
   

In [14]:
# create keyID
# GRCh37 v97 biomart
version37['stable_id2'] = version37['Gene stable ID'].astype(str) + \
'_' + version37['Transcript stable ID'].astype(str) + \
'_' + version37['Protein stable ID'].astype(str) + \
'_' + version37['CCDS ID'] + \
'_' + version37['UniProtKB/Swiss-Prot ID']

In [15]:
# GRCh38 v97 biomart
version38['stable_id2'] = version38['Gene stable ID'].astype(str) + \
'_' + version38['Transcript stable ID'].astype(str) + \
'_' + version38['Protein stable ID'].astype(str) + \
'_' + version38['CCDS ID'] + \
'_' + version38['UniProtKB/Swiss-Prot ID']


In [16]:
# get total # of unique UKB IDs
print(len(checkMatch.stable_id2.unique()))
print(len(version37.stable_id2.unique()))
print(len(version38.stable_id2.unique()))

# all the overlapping IDs between GRCh37 and GRCh38 are unique

22202
23334
38719


In [17]:
38719-22202
# diff between v97 38 and overlap len of 37 w/ 38

16517

In [18]:
23334-22202
# diff btw v97 37 with keyid2

1132

In [19]:
checkMatchlist = checkMatch['stable_id2'].tolist()
print(len(checkMatchlist))
print(checkMatchlist[0:10])

22202
['ENSG00000230031_ENST00000454856_ENSP00000456953_CCDS59248_H3BUK9', 'ENSG00000138593_ENST00000559471_ENSP00000453854_CCDS53942_Q93073', 'ENSG00000138593_ENST00000261847_ENSP00000261847_CCDS32234_Q93073', 'ENSG00000166351_ENST00000299443_ENSP00000299443_CCDS13562_Q86YR6', 'ENSG00000168675_ENST00000399848_ENSP00000382741_CCDS32794_O15165', 'ENSG00000188992_ENST00000344577_ENSP00000343331_CCDS13564_Q6XZB0', 'ENSG00000168675_ENST00000587757_ENSP00000466178_CCDS32795_O15165', 'ENSG00000168675_ENST00000586765_ENSP00000474783_CCDS42415_O15165', 'ENSG00000183706_ENST00000328795_ENSP00000332500_CCDS32173_Q8N0Y3', 'ENSG00000185272_ENST00000400577_ENSP00000383421_CCDS46635_P57052']


In [20]:
# label each 37 and 38 version file with True False 
# if stable_id is in the unique set of ID from my stable ID overlap between assemblies analysis
def addcolumnconditional(mapList, df, dfcol, newcol): 
    mendel = []
    for g in df[dfcol]:
        if g in mapList: 
            mendel.append("True")
        else:
            mendel.append("False")
    df.loc[:,newcol] = mendel
    return df

In [21]:
TF_37 = addcolumnconditional(checkMatchlist, version37, "stable_id2", "37match38stable_ids_2")


In [22]:
TF_38 = addcolumnconditional(checkMatchlist, version38, "stable_id2", "37match38stable_ids_2")


In [23]:
checkColumnValues(TF_37, "37match38stable_ids_2")

  37match38stable_ids_2  Count
0                  True  22202
1                 False   1132


In [24]:
checkColumnValues(TF_38, "37match38stable_ids_2")

  37match38stable_ids_2  Count
0                  True  22202
1                 False  16517


# biomart downloaded (CCDS filtered first) v97 files both equal 22,202 when filtered to match new KEY ID2...
### ENSG_ENST_ENSP_UKBSP_CCDS

### WITH only ensembl stable ID in key i had multiple mapping issue with keyID to unique UKBSP IDs



In [28]:
# drop all rows that have FALSE for match 
TF_37.drop(TF_37[TF_37["37match38stable_ids_2"] == "False"].index, inplace = True)
TF_38.drop(TF_38[TF_38["37match38stable_ids_2"] == "False"].index, inplace = True)

In [29]:
# check shape
describeMe(TF_37)
describeMe(TF_38)

(22202, 11)
Index(['Gene stable ID', 'Gene stable ID version', 'Transcript stable ID',
       'Transcript stable ID version', 'Protein stable ID',
       'Protein stable ID version', 'CCDS ID', 'UniProtKB/Swiss-Prot ID',
       'Transcript having exact match between ENSEMBL and HAVANA ID',
       'stable_id2', '37match38stable_ids_2'],
      dtype='object')
Gene stable ID                                                 0
Gene stable ID version                                         0
Transcript stable ID                                           0
Transcript stable ID version                                   0
Protein stable ID                                              0
Protein stable ID version                                      0
CCDS ID                                                        0
UniProtKB/Swiss-Prot ID                                        0
Transcript having exact match between ENSEMBL and HAVANA ID    0
stable_id2                                             

In [30]:
print("NUMBER of unique stable IDs")
print(len(TF_37.stable_id2.unique()))
print(len(TF_38.stable_id2.unique()))

NUMBER of unique stable IDs
22202
22202


In [31]:
print("NUMBER of unique gene stable IDs")
print(len(TF_37['Gene stable ID'].unique()))
print(len(TF_38['Gene stable ID'].unique()))

NUMBER of unique gene stable IDs
15837
15837


In [32]:
print("NUMBER of unique transcript stable IDs")
print(len(TF_37['Transcript stable ID'].unique()))
print(len(TF_38['Transcript stable ID'].unique()))

NUMBER of unique transcript stable IDs
22193
22193


In [33]:
print("NUMBER of unique protein stable IDs")
print(len(TF_37['Protein stable ID'].unique()))
print(len(TF_38['Protein stable ID'].unique()))

NUMBER of unique protein stable IDs
22193
22193


In [34]:
print("NUMBER of unique CCDS stable IDs")
print(len(TF_37['CCDS ID'].unique()))
print(len(TF_38['CCDS ID'].unique()))

NUMBER of unique CCDS stable IDs
21070
21070


In [35]:
print("NUMBER of unique Uniprot swissprot stable IDs")
print(len(TF_37['UniProtKB/Swiss-Prot ID'].unique()))
print(len(TF_38['UniProtKB/Swiss-Prot ID'].unique()))

NUMBER of unique Uniprot swissprot stable IDs
15819
15819


# total 22,202 unique stable_id2 overlapping 37 and 38 

### from GRCh37 v97 file with 23,334 rows & GRCh38 v97 file with 38,719 rows
> 22,202 match stableID overlap key

NUMBER of unique Uniprot swissprot stable IDs
> 15819


NUMBER of unique CCDS stable IDs
> 21070


NUMBER of unique protein stable IDs
> 22193


In [37]:
# saving
TF_37.to_csv("TF_37_22202_keyid_v2.csv",index=False)
TF_38.to_csv("TF_38_22202_keyid_v2.csv",index=False)

In [36]:
# importing entries from labeled ever file containing 3991 entries
labeledEver = pd.read_csv("/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CYS_LYS_2012_to2018UKBpositions/all3991_everLabeled_entries_simple.csv",sep='\t')
labeled = list(labeledEver.Entry)
print(len(labeled))


3991


In [38]:

# running laberler funx to filter for only rows that have UKBID also included in Nature experiment unique entries list
labeledTF37 = addcolumnconditional(labeled, TF_37,'UniProtKB/Swiss-Prot ID','Labeled_before_entry')
labeledTF38 = addcolumnconditional(labeled, TF_38,'UniProtKB/Swiss-Prot ID','Labeled_before_entry')

In [40]:
checkColumnValues(labeledTF37, 'Labeled_before_entry')
checkColumnValues(labeledTF38, 'Labeled_before_entry')

  Labeled_before_entry  Count
0                False  16841
1                 True   5361
  Labeled_before_entry  Count
0                False  16841
1                 True   5361


In [46]:
true37label = labeledTF37[labeledTF37['Labeled_before_entry'] == 'True']
true38label = labeledTF38[labeledTF38['Labeled_before_entry'] == 'True']

print(true37label.shape)
print(true38label.shape)

(5361, 12)
(5361, 11)


In [47]:
col2drop = "Transcript having exact match between ENSEMBL and HAVANA ID"

#DataFrame.drop(self, labels=None, axis=0, index=None, columns=None, level=None, inplace=False, errors='raise')[source]¶
# drop extra 37 col from biomart, HAVANA ID
true37label2 = true37label.drop(columns=[col2drop])


In [48]:
print(true37label2.shape)
print(true38label.shape)

(5361, 11)
(5361, 11)


In [49]:
# saving these new versions of GRCh37 and GRCh38 to make diff tables from
true37label2.to_csv("TF_37_truelabeled_before_5361_keyid2.csv", index=False)
true38label.to_csv("TF_38_truelabeled_before_5361_keyid2.csv",index=False)

In [5]:
# saving these new versions of GRCh37 and GRCh38 to make diff tables from
true37label2 = pd.read_csv("TF_37_truelabeled_before_5361_keyid2.csv") 
true38label = pd.read_csv("TF_38_truelabeled_before_5361_keyid2.csv")

In [9]:
# forgot to get unique counts
true37label2.head(3)

Gene stable ID Gene stable ID version Transcript stable ID  \
0  ENSG00000001497     ENSG00000001497.12      ENST00000374804   
1  ENSG00000001497     ENSG00000001497.12      ENST00000374807   
2  ENSG00000001497     ENSG00000001497.12      ENST00000374811   

  Transcript stable ID version Protein stable ID Protein stable ID version  \
0            ENST00000374804.5   ENSP00000363937         ENSP00000363937.5   
1            ENST00000374807.5   ENSP00000363940         ENSP00000363940.5   
2            ENST00000374811.3   ENSP00000363944         ENSP00000363944.3   

     CCDS ID UniProtKB/Swiss-Prot ID  \
0  CCDS55433                  Q9Y4W2   
1  CCDS55434                  Q9Y4W2   
2  CCDS14381                  Q9Y4W2   

                                          stable_id2  
0  ENSG00000001497_ENST00000374804_ENSP0000036393...  
1  ENSG00000001497_ENST00000374807_ENSP0000036394...  
2  ENSG00000001497_ENST00000374811_ENSP0000036394...

In [8]:
true38label.head(3)

Gene stable ID Gene stable ID version Transcript stable ID  \
0  ENSG00000001497     ENSG00000001497.16      ENST00000374804   
1  ENSG00000001497     ENSG00000001497.16      ENST00000374807   
2  ENSG00000001497     ENSG00000001497.16      ENST00000374811   

  Transcript stable ID version Protein stable ID Protein stable ID version  \
0            ENST00000374804.9   ENSP00000363937         ENSP00000363937.5   
1            ENST00000374807.9   ENSP00000363940         ENSP00000363940.5   
2            ENST00000374811.7   ENSP00000363944         ENSP00000363944.3   

     CCDS ID UniProtKB/Swiss-Prot ID  \
0  CCDS55433                  Q9Y4W2   
1  CCDS55434                  Q9Y4W2   
2  CCDS14381                  Q9Y4W2   

                                          stable_id2  
0  ENSG00000001497_ENST00000374804_ENSP0000036393...  
1  ENSG00000001497_ENST00000374807_ENSP0000036394...  
2  ENSG00000001497_ENST00000374811_ENSP0000036394...

In [10]:
print("NUMBER of unique stable IDs")
print(len(true37label2.stable_id2.unique()))
print(len(true38label.stable_id2.unique()))

print("NUMBER of unique gene stable IDs")
print(len(true37label2['Gene stable ID'].unique()))
print(len(true38label['Gene stable ID'].unique()))

print("NUMBER of unique transcript stable IDs")
print(len(true37label2['Transcript stable ID'].unique()))
print(len(true38label['Transcript stable ID'].unique()))

print("NUMBER of unique protein stable IDs")
print(len(true37label2['Protein stable ID'].unique()))
print(len(true38label['Protein stable ID'].unique()))

print("NUMBER of unique CCDS stable IDs")
print(len(true37label2['CCDS ID'].unique()))
print(len(true38label['CCDS ID'].unique()))

print("NUMBER of unique Uniprot swissprot stable IDs")
print(len(true37label2['UniProtKB/Swiss-Prot ID'].unique()))
print(len(true38label['UniProtKB/Swiss-Prot ID'].unique()))

NUMBER of unique stable IDs
5361
5361
NUMBER of unique gene stable IDs
3642
3642
NUMBER of unique transcript stable IDs
5361
5361
NUMBER of unique protein stable IDs
5361
5361
NUMBER of unique CCDS stable IDs
5049
5049
NUMBER of unique Uniprot swissprot stable IDs
3636
3636


In [ ]:
# added unique counts to powerpoint slides